base/REPL.jl

```diff
@@ -335,15 +345,30 @@ function hist_from_file(hp, file)
                 mode = symbol(m.captures[2])
             end
             line = hist_getline(file)
+            countlines += 1
         end
         isempty(line) && break
-        line[1] == '\t' || error(invalid_history_message)
+        # Make sure starts with tab
+        if line[1] == ' '
+            error(munged_history_message, countlines)
+        elseif line[1] != '\t'
+            error(invalid_history_message, hex(line[1]), " at line ", countlines)
+        end
         lines = UTF8String[]
         while !isempty(line)
             push!(lines, chomp(line[2:end]))
-            eof(file) && break
-            Base.peek(file) == '\t' || break
+            eof(file) &&
+                break
+            ch = Base.peek(file)
+            if ch == '#'
+                break
+            elseif ch == ' '
+                error(munged_history_message, countlines)
+            elseif ch != '\t'
+                error(invalid_history_message, hex(line[1]), " at line ", countlines)
+            end
```

Comment:
I suspect this somewhat repeated logic can be eliminated by checking the peek character and then letting it continue.